src/backends/backend/triton_model.cc

```diff
@@ -564,6 +564,67 @@ TRITONBACKEND_RequestRelease(
   return nullptr;  // success
 }
 
+///
+/// TRITONBACKEND_State
+///
+
+TRITONSERVER_Error*
+TRITONBACKEND_StateUpdate(TRITONBACKEND_State* state)
+{
+  InferenceRequest::State* ts =
+      reinterpret_cast<InferenceRequest::State*>(state);
+  auto status = ts->Update();
+
+  if (!status.IsOk()) {
+    return TRITONSERVER_ErrorNew(
+        StatusCodeToTritonCode(status.StatusCode()), status.Message().c_str());
+  }
+
+  return nullptr;  // success
+}
+
+TRITONSERVER_Error*
+TRITONBACKEND_StateNew(
+    TRITONBACKEND_State** state, TRITONBACKEND_Request* request,
+    const char* name, const TRITONSERVER_DataType datatype,
+    const int64_t* shape, const uint32_t dims_count)
+{
+  InferenceRequest* tr = reinterpret_cast<InferenceRequest*>(request);
+  std::vector<int64_t> lshape(shape, shape + dims_count);
+  InferenceRequest::State* lstate;
+  Status status =
+      tr->AddState(name, TritonToDataType(datatype), lshape, &lstate);
+  *lstate->MutableShape() = lshape;
+  if (!status.IsOk()) {
+    *state = nullptr;
+    return TRITONSERVER_ErrorNew(
+        StatusCodeToTritonCode(status.StatusCode()), status.Message().c_str());
+  }
+
+  *state = reinterpret_cast<TRITONBACKEND_State*>(lstate);
+  return nullptr;  // success
+}
+
+TRITONSERVER_Error*
+TRITONBACKEND_StateBuffer(
+    TRITONBACKEND_State* state, void** buffer, const uint64_t buffer_byte_size,
+    TRITONSERVER_MemoryType* memory_type, int64_t* memory_type_id)
+{
+  InferenceRequest::State* to =
+      reinterpret_cast<InferenceRequest::State*>(state);
+  std::shared_ptr<AllocatedMemory> memory = std::make_shared<AllocatedMemory>(
+      buffer_byte_size, *memory_type, *memory_type_id);
+  *buffer = memory->MutableBuffer(memory_type, memory_type_id);
```

Comment:
Shouldn't the state just have a persistent buffer that is used throughout the sequence lifetime... and only change when it needs to grow?

Reply:
In the current implementation it will be completely replaced with the new buffer. I'll modify it so that memory is allocated only when needed.